src/overlay/Peer.cpp

```diff
@@ -630,38 +710,256 @@ Peer::recvMessage(StellarMessage const& stellarMsg)
+         mtype = stellarMsg.type(), cat]() {
+            auto self = weak.lock();
+            if (!self)
             {
                 CLOG_TRACE(Overlay, "Error RecvMessage T:{} cat:{}", mtype,
                            cat);
+                return;
+            }
+
+            try
+            {
+                self->recvRawMessage(sm);
+            }
+            catch (CryptoError const& e)
+            {
+                std::string err = fmt::format(
+                    FMT_STRING("Error RecvMessage T:{} cat:{} {} @{:d}"), mtype,
+                    cat, self->toString(), app.getConfig().PEER_PORT);
+                CLOG_ERROR(Overlay, "Dropping connection with {}: {}", err,
+                           e.what());
+                self->drop("Bad crypto request",
+                           Peer::DropDirection::WE_DROPPED_REMOTE,
+                           Peer::DropMode::IGNORE_WRITE_QUEUE);
             }
+
+            // Notify on completion, schedule more reading if necessary
+            self->messageProcessed(sm);
         },
         fmt::format(FMT_STRING("{} recvMessage"), cat), type);
 }
 
+void
+Peer::sendSendMore(uint32_t numMessages)
+{
+    ZoneScoped;
+    StellarMessage m;
+    m.type(SEND_MORE);
+    m.sendMoreMessage().numMessages = numMessages;
+    sendMessage(m);
+}
+
+void
+Peer::recvSendMore(StellarMessage const& msg)
+{
+    ZoneScoped;
+    // Flow control must be "on"
+    assert(mRemoteOverlayVersion >=
+           Peer::FIRST_VERSION_SUPPORTING_FLOW_CONTROL);
+    assert(mRemotePeerWantsFlowControl && *mRemotePeerWantsFlowControl);
+    assert(msg.sendMoreMessage().numMessages > 0);
+
+    CLOG_TRACE(Overlay, "Peer {} sent SEND_MORE {}",
+               mApp.getConfig().toShortString(getPeerID()),
+               msg.sendMoreMessage().numMessages);
+
+    // SEND_MORE means we can free some capacity, and dump the next batch of
+    // messages onto the writing queue
+    mOutboundCapacity += msg.sendMoreMessage().numMessages;
+    maybeShedLoadAndSendNextBatch();
+}
+
+bool
+Peer::hasReadingCapacity() const
+{
+    bool enabledAndSupportedByRemotePeer =
+        (mRemoteOverlayVersion >= Peer::FIRST_VERSION_SUPPORTING_FLOW_CONTROL &&
+         mApp.getConfig().ENABLE_OVERLAY_FLOW_CONTROL);
+    return !enabledAndSupportedByRemotePeer || mCapacity.mTotalCapacity > 0;
+}
+
+static medida::Timer&
+getOutboundFloodTimer(Application& app, StellarMessage const& msg)
+{
+    std::string type;
+    switch (msg.type())
+    {
+    case SCP_MESSAGE:
+        type = "scp";
+        break;
+    case TRANSACTION:
+        type = "tx";
+        break;
+    default:
+        type = "ctrl";
+    }
+
+    return app.getMetrics().NewTimer({"overlay", "outbound-queue", type});
+}
+
+void
+Peer::maybeTrimQueue()
+{
+    auto now = mApp.getClock().now();
+    auto shouldTrim = [&](QueuedOutboundMessage const& msg) -> bool {
+        if (msg.mMessage.type() == TRANSACTION)
```

Comment:
we should just put an upper bound on the number of transactions that can sit there, that way you don't need to keep track of how long they wait and trim is a single call to "erase"